### PR TITLE
Climber

### PR DESCRIPTION
**Pull request by c0d3rman**
Added code for the climbing subsystem.

### CODE REVIEW

**Comment by c0d3rman** [comment]
This is the continuation of PR #8 - the branch was renamed.

**Comment by c0d3rman** [comment]
The latest commit is kinda janky because I wanted to keep the reference to the axis and the processing on NathanGain (to allow for multiple control schemes) but there's no way to pass an axis, just a thumbstick. I think we need to create an `AnalogSource` (and maybe `DigitalSource` and `RealSource`) which is just an interface with a getter that returns a value in the range `[-1, 1]` (and perhaps can accept a SpeedModifier?)

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java [hunk 1]
```diff
@@ -53,6 +56,21 @@ public void bindCommands() {
...
 		RobotMap.Component.driverXbox.dPad.right.whenReleased(normalDrive);
+		(new Climb(new PIDSource() { // Construct a PIDSource inline so we can process and pass the axis value here
+			@Override
+			public double pidGet() {
+				double climberSpeed = -RobotMap.Component.driverXbox.rightStick.getY(); // Inverted airplane-style analog control
+				return Math.max(0, climberSpeed); // Never drive the climber at a negative speed
+			}
+
+			@Override
+			public void setPIDSourceType(PIDSourceType pidSource) {}
```
Comment (leijurv) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java]: better to `throw new UnsupportedOperationException` than to return null and cause a null pointer exception somewhere else

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java [hunk 2]
```diff
@@ -53,6 +56,21 @@ public void bindCommands() {
 		RobotMap.Component.driverXbox.dPad.right.whenPressed(new ChassisTurnAbsolute(RobotMap.Component.chassis, 90,
 			RobotMap.Component.navx, RobotMap.Component.chassisTurnMC));
 		RobotMap.Component.driverXbox.dPad.right.whenReleased(normalDrive);
+		(new Climb(new PIDSource() { // Construct a PIDSource inline so we can process and pass the axis value here
```
Comment (leijurv) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java]: Outer parentheses not required. `new Climb(...).start();` is valid

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 3]
```diff
@@ -43,6 +44,10 @@
 			public static final int ballioDirectionalRoller = 1;
 			public static final int ballioHopperRollers = 2;
 			public static final int ballioElevatorAndIntakeRoller = 3;
+			public static final int climbMotorA = 4;
+			public static final int climbMotorB = 5;
+			public static final int flywheelLeftMotor = 6; // WIP
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/RobotMap.java:49]: Still WIP?

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 4]
```diff
@@ -109,8 +113,10 @@ public RobotMap() {
 			new EncoderGroup(100, Component.leftWheelEncoder, Component.rightWheelEncoder));
 		Component.leftWheel = new Motor("LeftWheel", false, new AccelerationCap(Component.pdp),
 			new VictorSP(Port.PWM.leftDriveA), new VictorSP(Port.PWM.leftDriveB));
+		Component.leftWheel.setInverted(true);
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/RobotMap.java:116]: Unrelated to climber -- still valid?

File: src/org/usfirst/frc4904/robot/commands/Climb.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+	public Climb(Supplier<Double> axis) {
+		super(RobotMap.Component.climber);
+		this.axis = axis;
+	}
+
+	@Override
+	public void execute() {
+		set(Math.max(0, axis.get())); // No, really, always feed positive values
+										// to the climber motor
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/Climb.java:19]: Is this really how the formatter wants it?

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/DefaultDriver.java [hunk 6]
```diff
@@ -1,5 +1,6 @@
 package org.usfirst.frc4904.robot.humaninterface.drivers;
 
+
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/DefaultDriver.java:3]: Formatter

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java [hunk 7]
```diff
@@ -53,6 +56,8 @@ public void bindCommands() {
 		RobotMap.Component.driverXbox.dPad.right.whenPressed(new ChassisTurnAbsolute(RobotMap.Component.chassis, 90,
 			RobotMap.Component.navx, RobotMap.Component.chassisTurnMC));
 		RobotMap.Component.driverXbox.dPad.right.whenReleased(normalDrive);
+		// Inverted airplane-style analog control
+		new Climb(() -> Math.max(0, -RobotMap.Component.driverXbox.rightStick.getY())).start();
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java:60]: I think the third level of safety here is unnecessary but not a blocker to merging.

File: src/org/usfirst/frc4904/sovereignty/TrimCommand.java [hunk 8]
```diff
@@ -1,5 +1,6 @@
 package org.usfirst.frc4904.sovereignty;
 
+
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/sovereignty/TrimCommand.java:3]: Formatter?

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 4 above)
Comment (ajnadel) [reply src/org/usfirst/frc4904/robot/RobotMap.java:116]: Yoni says this is valid.

**Review by ajnadel — APPROVED** [review]

**Comment by ajnadel** [comment]
Ignoring formatting changes because this is build season.

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 3 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/RobotMap.java:49]: The flywheel motors are. (No shooter yet)

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 4 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/RobotMap.java:116]: This is valid

File: src/org/usfirst/frc4904/robot/commands/Climb.java (hunk 5 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/commands/Climb.java:19]: I know, it's terrible. We're gonna revert the formatter line warp changes

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java (hunk 7 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java:60]: More safety is more betterer

**Pull request merged**